experiment/selinux/src/selinux.rs

```diff
@@ -0,0 +1,302 @@
+        Err(e) => return Err(e),
+    }
+}
+
+pub fn key_label() {
+    panic!("not implemented yet")
+}
+
+pub fn clear_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn reserve_label(label: &str) {
+    panic!("not implemented yet")
+}
+
+pub fn ro_file_label() {
+    panic!("not implemented yet")
+}
+
+pub fn kvm_container_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn init_container_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn container_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn priv_container_mount_label() {
+    panic!("not implemented yet")
+}
+
+pub fn format_mount_label(src: &str, mount_label: &str) -> String {
+    return format_mount_label_by_type(src, mount_label, "context")
+}
+
+pub fn format_mount_label_by_type(src: &str, mount_label: &str, context_type: &str) -> String {
+    let mut formatted_src = src.to_owned();
+
+    if !mount_label.is_empty() {
+        if formatted_src.is_empty() {
+            formatted_src = format!("{}=\"{}\"", context_type, mount_label);
+        } else {
+            formatted_src = format!("{},{}=\"{}\"", formatted_src, context_type, mount_label);
+        }
+    }
+    return formatted_src
+}
+
+pub fn write_con(fpath: &str, val: &str) -> Result<(), std::io::Error> {
+    panic!("not implemented yet");
+}
+
+pub fn is_proc_handle(file: &File) -> Result<(), std::io::Error> {
+    loop {
+        match statfs::fstatfs(file.as_fd()) {
+            Ok(stat) if stat.filesystem_type() == statfs::PROC_SUPER_MAGIC => break,
+            Ok(_) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other, format!("file {} is not on procfs", file.as_raw_fd())
+                ));
+            },
+            Err(Errno::EINTR) => continue,
+            Err(err) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("fstatfs failed: {}", err)))
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn read_con_fd(file: &mut File) -> Result<String, std::io::Error> {
+    let mut data = String::new();
+    file.read_to_string(&mut data)?;
+
+    let trimmed_data = data.trim_end_matches(char::from(0));
+    Ok(trimmed_data.to_string())
+}
+
+pub fn read_con(fpath: &str) -> Result<String, std::io::Error> {
+    if fpath.is_empty() {
+        return Err(std::io::Error::new(io::ErrorKind::InvalidInput, ERR_EMPTY_PATH));
+    }
+    let mut in_file = File::open(fpath)?;
+
+    is_proc_handle(&in_file)?;
+    return read_con_fd(&mut in_file);
+}
+
+pub fn attr_path(attr: &str) -> String {
+    const THREAD_SELF_PREFIX: &str = "/proc/thread-self/attr";
+    ATTR_PATH_INIT.call_once(|| {
+        let path = PathBuf::from(THREAD_SELF_PREFIX);
+        unsafe {
+            HAVE_THREAD_SELF = path.is_dir();
+        }
+    });
+
+    unsafe {
+        if HAVE_THREAD_SELF {
+            return format!("{}/{}", THREAD_SELF_PREFIX, attr);
+        }
+    }
+
```

Comment:
Can you explain the use of this function? I am not able to understand it correctly :sweat_smile:

1. What is the reason for using Once and call_once here? Assuming youki as the primary use-case, is there still a chance of having a multi-thread race? If so can be use AtomicBool instead and remove dependency of Once? (Suggesting as underlying data is simple bool type). We might  be able to do an compare-and-exchange operation instead of the call_once here.
2. Do we need the `unsafe` block if we are only reading the `HAVE_THREAD_SELF` value? I thought unsafe was needed only if we were writing to the variable.

Reply:
Thanks, I've updated the code using atomicBool instead.

I implemented using Once because go-selinux implemented like that.
I guess that this is to avoid conflicts between multi-thread like goroutine.
I thought it is possible that similar issue happens on Youki as well.
https://github.com/opencontainers/selinux/blob/9dee859cbffb974e779f65efaef9f5ef6e1d260b/go-selinux/selinux_linux.go#L451-L467

I am not familiar with selinux nor youki, please let me know if I misunderstood.

Reply:
Hey, thanks. I am also not that familiar with selinux-go, so will take a look and confirm once.